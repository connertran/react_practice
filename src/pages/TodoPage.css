.todo-list {
  display: flex;
  flex-direction: column;
  border: solid black 2px;
}
