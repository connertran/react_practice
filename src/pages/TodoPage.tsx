import React from "react";
import { useState, useCallback } from "react";
import TodoList from "../todo/TodoList";
import TodoAddBar from "../todo/TodoAddBar";
import { Todo } from "../types/todo";
import { initialTodos } from "../data/initialTodos";

function TodoPage() {
  const [todoState, setTodoState] = useState<Todo[]>(initialTodos);
  const [showError, setShowError] = useState<string>("");

  const handleFinishTodo = useCallback(
    (id: number): void => {
      // find the todo in the array
      let index = -1;
      for (let i = 0; i < todoState.length; i++) {
        if (todoState[i].id === id) {
          index = i;
          break;
        }
      }
      if (index === -1) {
        setShowError("Cannot find the todo. Refresh the page and try again.");
        return;
      }

      //change the boolean
      setTodoState((prev) => {
        const next = [...prev];
        next[index] = { ...next[index], finished: !next[index].finished };
        return next;
      });
    },
    [todoState]
  );
  return (
    <div>
      {showError === "" ? <p>{showError}</p> : null}

      <h1>Todo List</h1>
      <TodoAddBar setTodoState={setTodoState} />
      <ul className="todo-list">
        <TodoList
          setTodoState={setTodoState}
          todos={todoState}
          toggleTodo={handleFinishTodo}
          displayError={setShowError}
        />
      </ul>
    </div>
  );
}

export default TodoPage;
