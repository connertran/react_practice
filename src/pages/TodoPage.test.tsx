import { render, fireEvent, screen } from "@testing-library/react";
import TodoPage from "./TodoPage";
describe("Todo Page", () => {
  it("it renders without breaking", function () {
    render(<TodoPage />);
  });

  // snapshot test
  it("matches the previous snapshot", () => {
    const { asFragment } = render(<TodoPage />);
    expect(asFragment()).toMatchSnapshot();
  });

  // test deleting todo
  it("delete button deletes the todo", () => {
    const { getByText, getAllByText } = render(<TodoPage />);
    expect(getByText("Create a Todo React app")).toBeInTheDocument();
    const buttonsArr = getAllByText("X");
    const firstButton = buttonsArr[0];
    fireEvent.click(firstButton);
    expect(() => {
      getByText("Create a Todo React app");
    }).toThrow();
  });
});
