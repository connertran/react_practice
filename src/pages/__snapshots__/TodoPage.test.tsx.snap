// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Page matches the previous snapshot 1`] = `
<DocumentFragment>
  <div>
    <p />
    <h1>
      Todo List
    </h1>
    <form>
      <input
        placeholder="Add your todo"
        type="text"
        value=""
      />
      <button>
        Add
      </button>
    </form>
    <ul
      class="todo-list"
    >
      <li>
        <label>
          Create a Todo React app
        </label>
        <input
          type="checkbox"
        />
        <button>
          X
        </button>
      </li>
      <li>
        <label>
          Exercise
        </label>
        <input
          checked=""
          type="checkbox"
        />
        <button>
          X
        </button>
      </li>
      <li>
        <label>
          Interview
        </label>
        <input
          checked=""
          type="checkbox"
        />
        <button>
          X
        </button>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
